File: javascript/package_artifacts.js
#!/usr/bin/env node
// package_artifacts.py — portiert nach javascript
// Quelle: python, Projects@TikTok-Live-Companion:plugin-source/scripts/package_artifacts.py
// auch in: Projects@TikTok-Live-Companion-Android:plugin-source/scripts/package_artifacts.py
// auch in: Projects@TikTok-Live-Companion-iOS:plugin-source/scripts/package_artifacts.py
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const yargs = require('yargs');
const AdmZip = require('adm-zip');

const __filename = process.argv[1];
const ROOT = path.resolve(path.dirname(__filename), '../..');
const PROJECT_ROOT = path.resolve(ROOT, '..');
const EXCLUDED_PARTS = new Set(["__pycache__", ".gradle", ".kotlin", "build", "DerivedData", "xcuserdata"]);

function addTree(archive, source, prefix = "") {
    const files = getAllFiles(source);
    files.sort();
    for (const filePath of files) {
        const relativePath = path.relative(source, filePath);
        const parts = relativePath.split(path.sep);
        if (EXCLUDED_PARTS.has(parts[0]) || path.extname(filePath) === '.pyc' || path.extname(filePath) === '.aar') {
            continue;
        }
        const archivePath = path.posix.join(prefix, relativePath);
        const fileContent = fs.readFileSync(filePath);
        archive.addFile(archivePath, fileContent);
    }
}

function getAllFiles(dir) {
    let results = [];
    const list = fs.readdirSync(dir);
    list.forEach(file => {
        file = path.resolve(dir, file);
        const stat = fs.statSync(file);
        if (stat && stat.isDirectory()) {
            results = [...results, ...getAllFiles(file)];
        } else {
            results.push(file);
        }
    });
    return results;
}

const argv = yargs
    .usage('Usage: $0 [options]')
    .describe('output-dir', 'Output directory')
    .demandOption(['output-dir'])
    .option('android-apk', {
        describe: 'Optional verified mockDebug or shazamDebug APK',
        type: 'string'
    })
    .option('android-source', {
        describe: 'Verified Android source root',
        default: path.join(PROJECT_ROOT, 'mobile', 'android'),
        type: 'string'
    })
    .option('ios-source', {
        describe: 'Verified iOS source root',
        default: path.join(PROJECT_ROOT, 'mobile', 'ios'),
        type: 'string'
    })
    .argv;

const outputDir = path.resolve(argv.outputDir);
fs.mkdirSync(outputDir, { recursive: true });

const manifestPath = path.join(ROOT, 'browser-extension', 'manifest.json');
const manifest = JSON.parse(fs.readFileSync(manifestPath, 'utf-8'));
const version = manifest.version;

const extensionZip = path.join(outputDir, `tiktok-live-companion-extension-${version}.zip`);
const pluginZip = path.join(outputDir, `tiktok-live-companion-plugin-${version}.zip`);
const serviceZip = path.join(outputDir, `tiktok-live-companion-service-${version}.zip`);
const iosSourceZip = path.join(outputDir, `tiktok-live-companion-ios-${version}-source.zip`);
const androidSourceZip = path.join(outputDir, `tiktok-live-companion-android-${version}-source.zip`);
const androidApk = path.join(outputDir, `tiktok-live-companion-android-${version}.apk`);
const extensionDir = path.join(outputDir, `tiktok-live-companion-extension-${version}`);
const checksumFile = path.join(outputDir, `tiktok-live-companion-${version}-SHA256.txt`);

const resolvedExtensionDir = path.resolve(extensionDir);
if (path.dirname(resolvedExtensionDir) !== outputDir) {
    throw new Error("Refusing to package outside the requested output directory");
}
if (fs.existsSync(extensionDir)) {
    fs.rmSync(extensionDir, { recursive: true });
}
copyDirSync(path.join(ROOT, 'browser-extension'), extensionDir);
copyDirSync(path.join(ROOT, 'companion-service'), path.join(extensionDir, 'companion-service'));

const batchScriptContent = '@echo off\r\ncall "%~dp0companion-service\\Sprachdienst-reparieren.cmd"\r\n';
fs.writeFileSync(path.join(extensionDir, 'Sprachdienst-reparieren.cmd'), batchScriptContent, 'utf-8');

const packageJsonContent = {
    name: 'tiktok-live-companion-extension-package',
    private: true,
    version: version,
    scripts: {
        setup: 'npm --prefix companion-service run setup --',
        start: 'npm --prefix companion-service start',
        test: 'npm --prefix companion-service test'
    }
};
fs.writeFileSync(path.join(extensionDir, 'package.json'), JSON.stringify(packageJsonContent, null, 2) + '\n', 'utf-8');

let archive = new AdmZip();
addTree(archive, extensionDir);
archive.writeZip(extensionZip);

archive = new AdmZip();
addTree(archive, ROOT, 'tiktok-live-companion');
archive.writeZip(pluginZip);

archive = new AdmZip();
addTree(archive, path.join(ROOT, 'companion-service'));
archive.writeZip(serviceZip);

const iosSource = path.resolve(argv.iosSource);
const androidSource = path.resolve(argv.androidSource);
if (!fs.existsSync(iosSource) || !fs.statSync(iosSource).isDirectory() ||
    !fs.existsSync(androidSource) || !fs.statSync(androidSource).isDirectory()) {
    throw new Error("--ios-source and --android-source must point to existing source directories");
}

archive = new AdmZip();
addTree(archive, iosSource, 'TikTokLiveCompanion-iOS');
archive.writeZip(iosSourceZip);

archive = new AdmZip();
addTree(archive, androidSource, 'TikTokLiveCompanion-Android');
archive.writeZip(androidSourceZip);

if (argv.androidApk) {
    const sourceApk = path.resolve(argv.androidApk);
    if (!fs.existsSync(sourceApk) || path.extname(sourceApk).toLowerCase() !== '.apk') {
        throw new Error("--android-apk must point to an existing APK");
    }
    if (path.resolve(sourceApk) !== path.resolve(androidApk)) {
        fs.copyFileSync(sourceApk, androidApk);
    }
}

let artifacts = [extensionZip, pluginZip, serviceZip, iosSourceZip, androidSourceZip];
if (fs.existsSync(androidApk)) {
    artifacts.push(androidApk);
}
const checksums = [];
for (const artifact of artifacts) {
    const fileBuffer = fs.readFileSync(artifact);
    const hashSum = crypto.createHash('sha256');
    hashSum.update(fileBuffer);
    const hex = hashSum.digest('hex');
    checksums.push(`${hex}  ${path.basename(artifact)}`);
}
fs.writeFileSync(checksumFile, checksums.join('\n') + '\n', 'utf-8');

console.log(JSON.stringify({
    extension_dir: extensionDir,
    extension_zip: extensionZip,
    plugin_zip: pluginZip,
    service_zip: serviceZip,
    ios_source_zip: iosSourceZip,
    android_source_zip: androidSourceZip,
    android_apk: fs.existsSync(androidApk) ? androidApk : null,
    checksum_file: checksumFile,
    version: version
}, null, 2));

function copyDirSync(src, dest) {
    const entries = fs.readdirSync(src, { withFileTypes: true });
    fs.mkdirSync(dest, { recursive: true });
    for (let entry of entries) {
        const srcPath = path.join(src, entry.name);
        const destPath = path.join(dest, entry.name);
        if (entry.isDirectory()) {
            copyDirSync(srcPath, destPath);
        } else {
            fs.copyFileSync(srcPath, destPath);
        }
    }
}

File: perl5/package_artifacts.pl
#!/usr/bin/perl
# package_artifacts.py — portiert nach perl5
# Quelle: python, Projects@TikTok-Live-Companion:plugin-source/scripts/package_artifacts.py
# auch in: Projects@TikTok-Live-Companion-Android:plugin-source/scripts/package_artifacts.py
# auch in: Projects@TikTok-Live-Companion-iOS:plugin-source/scripts/package_artifacts.py
# Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use Getopt::Long;
use JSON;
use File::Path qw(make_path remove_tree);
use File::Copy qw(copy);
use File::Spec;
use File::Find;
use Digest::SHA qw(sha256_hex);
use Archive::Zip qw(:CONSTANTS :ERROR_CODES);

my $ROOT = do {
    my @parts = split '/', __FILE__;
    pop @parts;
    pop @parts;
    join('/', @parts);
};

my $PROJECT_ROOT = do {
    my @parts = split '/', $ROOT;
    pop @parts;
    join('/', @parts);
};

my %EXCLUDED_PARTS = map { $_ => 1 } ("__pycache__", ".gradle", ".kotlin", "build", "DerivedData", "xcuserdata");

sub add_tree {
    my ($archive, $source, $prefix) = @_;
    $prefix //= "";
    
    my @files;
    find(sub {
        return unless -f $_;
        my $rel_path = substr($File::Find::name, length($source) + 1);
        my @parts = split '/', $rel_path;
        
        # Check excluded parts
        my $excluded = 0;
        for my $part (@parts) {
            if ($EXCLUDED_PARTS{$part}) {
                $excluded = 1;
                last;
            }
        }
        
        # Check file extensions
        my ($ext) = ($_ =~ /(\.[^.]+)$/);
        if (!$excluded && $ext && ($ext eq '.pyc' || $ext eq '.aar')) {
            $excluded = 1;
        }
        
        push @files, $File::Find::name unless $excluded;
    }, $source);
    
    @files = sort @files;
    
    for my $file (@files) {
        my $relative = substr($file, length($source) + 1);
        my $archive_path = $prefix ? "$prefix/$relative" : $relative;
        $archive_path =~ s|\\|/|g;  # Normalize path separators
        
        my $member = $archive->addFile($file, $archive_path);
        if ($member) {
            $member->desiredCompressionMethod(COMPRESSION_DEFLATED);
            $member->unixFileAttributes(0100644);
        }
    }
}

# Parse arguments
my $output_dir;
my $android_apk;
my $android_source = "$PROJECT_ROOT/mobile/android";
my $ios_source = "$PROJECT_ROOT/mobile/ios";

GetOptions(
    "output-dir=s" => \$output_dir,
    "android-apk=s" => \$android_apk,
    "android-source=s" => \$android_source,
    "ios-source=s" => \$ios_source,
) or die "Invalid options\n";

die "--output-dir is required\n" unless $output_dir;

# Create output directory
make_path($output_dir) unless -d $output_dir;

# Resolve paths
$output_dir = File::Spec->rel2abs($output_dir);
$android_source = File::Spec->rel2abs($android_source);
$ios_source = File::Spec->rel2abs($ios_source);
$android_apk = File::Spec->rel2abs($android_apk) if $android_apk;

# Read manifest
my $manifest_content;
{
    open my $fh, '<:encoding(UTF-8)', "$ROOT/browser-extension/manifest.json" or die "Cannot read manifest.json: $!";
    local $/;
    $manifest_content = <$fh>;
    close $fh;
}
my $manifest = decode_json($manifest_content);
my $version = $manifest->{version};

# Define output files
my $extension_zip = "$output_dir/tiktok-live-companion-extension-$version.zip";
my $plugin_zip = "$output_dir/tiktok-live-companion-plugin-$version.zip";
my $service_zip = "$output_dir/tiktok-live-companion-service-$version.zip";
my $ios_source_zip = "$output_dir/tiktok-live-companion-ios-$version-source.zip";
my $android_source_zip = "$output_dir/tiktok-live-companion-android-$version-source.zip";
my $android_apk_out = "$output_dir/tiktok-live-companion-android-$version.apk";
my $extension_dir = "$output_dir/tiktok-live-companion-extension-$version";
my $checksum_file = "$output_dir/tiktok-live-companion-$version-SHA256.txt";

# Validate extension directory location
my $resolved_extension_dir = File::Spec->rel2abs($extension_dir);
my ($volume, $directories, $file) = File::Spec->splitpath($resolved_extension_dir);
my @dirs = grep { $_ ne '' } File::Spec->splitdir($directories);
pop @dirs;  # Remove the last directory (the extension dir name)
my $parent_dir = File::Spec->catpath($volume, File::Spec->catdir(@dirs), '');
if ($parent_dir ne $output_dir) {
    die "Refusing to package outside the requested output directory\n";
}

# Clean and copy extension directory
remove_tree($extension_dir) if -d $extension_dir;
make_path("$extension_dir/companion-service");

# Copy browser-extension contents
opendir(my $dh, "$ROOT/browser-extension") or die "Cannot open directory: $!";
my @entries = readdir($dh);
closedir($dh);
for my $entry (@entries) {
    next if $entry eq '.' || $entry eq '..';
    my $src = "$ROOT/browser-extension/$entry";
    my $dst = "$extension_dir/$entry";
    if (-d $src) {
        system("cp", "-r", $src, $dst) == 0 or die "Copy failed: $!";
    } else {
        copy($src, $dst) or die "Copy failed: $!";
    }
}

# Copy companion-service contents
opendir($dh, "$ROOT/companion-service") or die "Cannot open directory: $!";
@entries = readdir($dh);
closedir($dh);
for my $entry (@entries) {
    next if $entry eq '.' || $entry eq '..';
    my $src = "$ROOT/companion-service/$entry";
    my $dst = "$extension_dir/companion-service/$entry";
    if (-d $src) {
        system("cp", "-r", $src, $dst) == 0 or die "Copy failed: $!";
    } else {
        copy($src, $dst) or die "Copy failed: $!";
    }
}

# Write Windows batch file
{
    open my $fh, '>:encoding(UTF-8)', "$extension_dir/Sprachdienst-reparieren.cmd" or die "Cannot write batch file: $!";
    print $fh '@echo off' . "\r\n" . 'call "%~dp0companion-service\Sprachdienst-reparieren.cmd"' . "\r\n";
    close $fh;
}

# Write package.json
my $package_json = {
    name => "tiktok-live-companion-extension-package",
    private => JSON::true,
    version => $version,
    scripts => {
        setup => "npm --prefix companion-service run setup --",
        start => "npm --prefix companion-service start",
        test => "npm --prefix companion-service test"
    }
};
{
    open my $fh, '>:encoding(UTF-8)', "$extension_dir/package.json" or die "Cannot write package.json: $!";
    print $fh to_json($package_json, { utf8 => 1, pretty => 1 }) . "\n";
    close $fh;
}

# Create extension zip
{
    my $zip = Archive::Zip->new();
    add_tree($zip, $extension_dir);
    die "Write error\n" unless $zip->writeToFileNamed($extension_zip) == AZ_OK;
}

# Create plugin zip
{
    my $zip = Archive::Zip->new();
    add_tree($zip, $ROOT, "tiktok-live-companion");
    die "Write error\n" unless $zip->writeToFileNamed($plugin_zip) == AZ_OK;
}

# Create service zip
{
    my $zip = Archive::Zip->new();
    add_tree($zip, "$ROOT/companion-service");
    die "Write error\n" unless $zip->writeToFileNamed($service_zip) == AZ_OK;
}

# Validate source directories
die "--ios-source and --android-source must point to existing source directories\n" 
    unless (-d $ios_source && -d $android_source);

# Create iOS source zip
{
    my $zip = Archive::Zip->new();
    add_tree($zip, $ios_source, "TikTokLiveCompanion-iOS");
    die "Write error\n" unless $zip->writeToFileNamed($ios_source_zip) == AZ_OK;
}

# Create Android source zip
{
    my $zip = Archive::Zip->new();
    add_tree($zip, $android_source, "TikTokLiveCompanion-Android");
    die "Write error\n" unless $zip->writeToFileNamed($android_source_zip) == AZ_OK;
}

# Handle Android APK
if ($android_apk) {
    die "--android-apk must point to an existing APK\n" 
        unless (-f $android_apk && $android_apk =~ /\.apk$/i);
    
    if (File::Spec->rel2abs($android_apk) ne File::Spec->rel2abs($android_apk_out)) {
        copy($android_apk, $android_apk_out) or die "Failed to copy APK: $!";
    }
}

# Calculate checksums
my @artifacts = ($extension_zip, $plugin_zip, $service_zip, $ios_source_zip, $android_source_zip);
push @artifacts, $android_apk_out if -f $android_apk_out;

my @checksums;
for my $artifact (@artifacts) {
    open my $fh, '<', $artifact or die "Cannot read $artifact: $!";
    binmode $fh;
    my $digest = sha256_hex(do { local $/; <$fh> });
    close $fh;
    my ($name) = ($artifact =~ /([^\/\\]+)$/);
    push @checksums, "$digest  $name";
}

{
    open my $fh, '>:encoding(UTF-8)', $checksum_file or die "Cannot write checksum file: $!";
    print $fh join("\n", @checksums) . "\n";
    close $fh;
}

# Output result JSON
my $result = {
    extension_dir => File::Spec->rel2abs($extension_dir),
    extension_zip => File::Spec->rel2abs($extension_zip),
    plugin_zip => File::Spec->rel2abs($plugin_zip),
    service_zip => File::Spec->rel2abs($service_zip),
    ios_source_zip => File::Spec->rel2abs($ios_source_zip),
    android_source_zip => File::Spec->rel2abs($android_source_zip),
    android_apk => (-f $android_apk_out) ? File::Spec->rel2abs($android_apk_out) : undef,
    checksum_file => File::Spec->rel2abs($checksum_file),
    version => $version
};

print encode_json($result) . "\n";
